### PR TITLE
cli,sdnotify: improve the handling of unix socket directories

### LINKED ISSUE #76904: cli/start: sdnotify socket listen chokes if $PWD is too long
We create cockroach single node instances in tests, to write integration-style tests against. Our builds happen in sandboxes; in particular TMPDIR is typically set to a very long path.

We don't want to hit disk so we point the cockroach node at a ram disk. We pass --temp-dir and --socket-dir ; something in cockroach ignores these and attempts to create a unix socket in $TMPDIR. This fails because the maximum length of a unix socket name is 108 chars, and our sandbox's tmp dir name takes the length over that.

Example, doing things straight on the command line. (This is on EL 7.)

Set up directories:

```
~ $ mkdir $HOME/this-is-a-very-long-directory-name-the-point-is-to-be-more-than-one-hundred-and-eight-characters

~ $ export TMPDIR=$HOME/this-is-a-very-long-directory-name-the-point-is-to-be-more-than-one-hundred-and-eight-characters

~ $ mkdir ~/local/sensible-length-path

~ $ export BASE=$HOME/sensible-length-path

~ $ mkdir -p $BASE/{tmp,store,socket}

Try to start cockroach

~ $ cockroach version
Build Tag:        v21.2.5
Build Time:       2022/02/07 21:01:07
Distribution:     CCL
Platform:         linux amd64 (x86_64-unknown-linux-gnu)
Go Version:       go1.16.6
C Compiler:       gcc 6.5.0
Build Commit ID:  5afb632f77eee9f09f2adfa2943e1979ec4ebedf
Build Type:       release

~ $ cockroach start-single-node --insecure --background --listen-addr=localhost:9090 --http-addr=localhost:9091 --pid-file=$BASE/pid --store=$BASE/store --temp-dir=$BASE/tmp --socket-dir=$BASE/socket

E220211 05:05:00.202740 1 1@cli/clierror/check.go:35  [-] 1  ERROR: cannot dial server.
E220211 05:05:00.202740 1 1@cli/clierror/check.go:35  [-] 1 +Is the server running?
E220211 05:05:00.202740 1 1@cli/clierror/check.go:35  [-] 1 +If the server is running, check --host client-side and --advertise server-side.
E220211 05:05:00.202740 1 1@cli/clierror/check.go:35  [-] 1 +listen unixgram /home/foo/this-is-a-very-long-directory-name-the-point-is-to-be-more-than-one-hundred-and-twenty-three-characters/sdnotify058488223/notify.sock: bind: invalid argument
ERROR: cannot dial server.
Is the server running?
If the server is running, check --host client-side and --advertise server-side.

listen unixgram /home/foo/this-is-a-very-long-directory-name-the-point-is-to-be-more-than-one-hundred-and-twenty-three-characters/sdnotify058488223/notify.sock: bind: invalid argument
Failed running "start-single-node"
EXIT STATUS 1
```

Override TMPDIR to something reasonable and cockroach works:

```
~ $ export TMPDIR=$BASE/tmp

~ $ cockroach start-single-node --insecure --background --listen-addr=localhost:9090 --http-addr=localhost:9091 --pid-file=$BASE/pid --store=$BASE/store --temp-dir=$BASE/tmp --socket-dir=$BASE/socket
*
* WARNING: ALL SECURITY CONTROLS HAVE BEEN DISABLED!
*
* This mode is intended for non-production testing only.
*
* In this mode:
* - Your cluster is open to any client that can access localhost.
* - Intruders with access to your machine or network can observe client-server traffic.
* - Intruders can log in without password and read or write any data in the cluster.
* - Intruders can consume all your server's resources and cause unavailability.
*
*
* INFO: To start a secure server without mandating TLS for clients,
* consider --accept-sql-without-tls instead. For other options, see:
*
* - https://go.crdb.dev/issue-v/53404/v21.2
* - https://www.cockroachlabs.com/docs/v21.2/secure-a-cluster.html
*
```

This looks like something internal is preferring env variables to explicitly passed flags, which I think is a bug.

Jira issue: CRDB-13317

### PR DESCRIPTION
**Pull request by knz**
tldr: This patch makes it easier for users to understand when
CockroachDB has a problem using/creating unix sockets, and
what they can do to avert the situation.

Fixes #76904.

Example, before;
```
$ cockroach start-single-node --insecure --background
...
listen unixgram /data/home/kena/cockroach/this-is-a-very-long-directory-name-the-point-is-to-be-more-than-one-hundred-and-twenty-three-characters/sdnotify2139543236/notify.sock: bind: invalid argument
```

Example, after:
```
$ cockroach start-single-node --insecure --background
ERROR: no suitable directory found for the --background notify socket
HINT: Avoid using --background altogether (preferred), or use a
shorter directory name for --socket-dir, TMPDIR or the current directory.
```

**Context:**

CockroachDB uses unix sockets both to handle `--background` and to
create a postgres-compatible SQL client socket via `--socket-dir`. For
`--background`, the logic was previously hardwired to always use
`os.TempDir()`, i.e. either `$TMPDIR` or `/tmp`.

On Unix, socket objects are commonly limited to 128 bytes;
that's e.g. 104 path characters including final NUL on BSD, 108 on glibc.
When the actual length is larger, creating the socket fails with
a confusing error `bind: invalid argument`.

**The problem:**

Users have reported issues in the following circumstances:

- when setting their CWD/$TMPDIR to a long directory name,
  they expected `--socket-dir` to also influence the socket
  name used for `--background`. Instead, `--background`
  failed with a hard-to-understand error. (#76904)

- when running tests inside a container/sandbox with a long
  prefix, they were confused that `--background` / `--socket-dir=.`
  did not work properly.

- they use macOS which has very large directory names in $TMPDIR.

To improve user experience, this patch changes the behavior
as described in the release notes below below.

Release note (cli change): CockroachDB now produces a clearer error
when the path specified via `--socket-dir` is too long.

Release note (cli change): When the flag `--background` is specified,
CockroachDB now makes 3 attempts to find a suitable directory to
create the notification socket: the value of `--socket-dir` if
specified, the value of `$TMPDIR` (or /tmp if the env var is empty),
and the current working directory. If none of these directories has a
name short enough, an explanatory error is printed.

### CODE REVIEW

**Review by jreut — APPROVED** [review]
LGTM. I see mention of long macOS $TMP paths. Have these tests simply historically failed on macOS?

> Reviewed 7 of 7 files at r1, 1 of 1 files at r2, 3 of 3 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/84532#-:-N7MkujF5trtx_Qf90T1:bxalr8t)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @bdarnell and @stevendanna)

___
*[`pkg/cli/flags.go` line 1013 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/84532#-N7MjTFVAQ52afydbgKa:-N7MjTFVAQ52afydbgKb:b8ycu70) ([raw file](https://github.com/cockroachdb/cockroach/blob/265667cab6fa3c98548d58f4f87521e4d53b7541/pkg/cli/flags.go#L1013)):*
> ```Go
> 		} else {
> 			socketName := ".s.PGSQL." + serverListenPort
> 			// On BSD, binding to a socket is limited to a path length of 104 characters
> ```

For my own education, could you link to the code or docs that support these facts?
___
*[`pkg/cli/interactive_tests/test_socket_name.tcl` line 65 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/84532#-N7Mk4sKDU3D3eqzKnG4:-N7Mk4sKDU3D3eqzKnG5:bd6njic) ([raw file](https://github.com/cockroachdb/cockroach/blob/265667cab6fa3c98548d58f4f87521e4d53b7541/pkg/cli/interactive_tests/test_socket_name.tcl#L65)):*
> ```Tcl
> system "$crdb sql --insecure -e 'select 1'"
> stop_server $crdb
> end_test
> ```

TIL we have TCL-based tests in CRDB. These are a joy to read!

**Review by knz — COMMENTED** [review]
> Have these tests simply historically failed on macOS?

They do when you try to run them outside of the docker sandbox.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/84532#-:-N7Mman364nrMwAf3mT3:b1g3qvx)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @bdarnell and @stevendanna)

___
*[`pkg/cli/flags.go` line 1013 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/84532#-N7MjTFVAQ52afydbgKa:-N7Mm2QE4y9pmNVSCqgE:bco51qi) ([raw file](https://github.com/cockroachdb/cockroach/blob/265667cab6fa3c98548d58f4f87521e4d53b7541/pkg/cli/flags.go#L1013)):*
<details><summary><i>Previously, jreut (jordan ryan reuter) wrote…</i></summary><blockquote>

For my own education, could you link to the code or docs that support these facts?
</blockquote></details>

https://www.man7.org/linux/man-pages/man7/unix.7.html
```
      A UNIX domain socket address is represented in the following
       structure:

           struct sockaddr_un {
               sa_family_t sun_family;               /* AF_UNIX */
               char        sun_path[108];            /* Pathname */
           };

       The sun_family field always contains AF_UNIX.  On Linux, sun_path
       is 108 bytes in size
```

https://www.freebsd.org/cgi/man.cgi?query=unix&sektion=4
```
UNIX-domain addresses are variable-length file system pathnames of	at
     most 104 characters.  The include file <sys/un.h> defines this address:

	   struct sockaddr_un {
		   u_char  sun_len;
		   u_char  sun_family;
		   char	   sun_path[104];
	   };
```

**Comment by knz** [comment]
TFYR!

bors r=jreut
